File: bindings/bun/demo/02_precision.ts
/**
 * Copyright (c) 2026 Edison Manrique Chocce
 * Licencia Dual: AGPL-3.0 | Comercial según LICENSE.
 *
 * demo/02_precision.ts — Aritmética Compensada (Neumaier)
 */
import { Precision } from "../src/index";

console.log("=".repeat(60));
console.log("  ERMC | Demo 02: Aritmética Compensada (Neumaier)");
console.log("=".repeat(60));

// [1] Cancelación catastrófica
const data = [1e16, 1.0, -1e16];
const naiveSum = data.reduce((a, b) => a + b, 0);
const zigSum = Precision.sum(data);
console.log("\n[1] Cancelación catastrófica clásica:");
console.log(`  Datos: [1e16, 1.0, -1e16]`);
console.log(`  JS reduce():    ${naiveSum}   <-- ERROR: 0.0`);
console.log(`  Neumaier Zig:  ${zigSum}  <-- CORRECTO: 1.0`);

// [2] Serie armónica H_N
const N = 10_000;
const harmonic = Array.from({ length: N }, (_, k) => 1.0 / (k + 1));
const jsSumH = harmonic.reduce((a, b) => a + b, 0);
const zigSumH = Precision.sum(harmonic);
const ref = Math.log(N) + 0.5772156649015328 + 1 / (2 * N);
console.log(`\n[2] Serie armónica H_${N}:`);
console.log(`  JS reduce():  ${jsSumH.toFixed(10)}`);
console.log(`  Neumaier Zig: ${zigSumH.toFixed(10)}`);
console.log(`  Referencia:   ${ref.toFixed(10)}`);
console.log(`  Error JS:     ${Math.abs(jsSumH - ref).toExponential(2)}`);
console.log(`  Error Zig:    ${Math.abs(zigSumH - ref).toExponential(2)}`);

// [3] Magnitudes dispares
const dispares = [1e-15, 1e15, 1e-15, -1e15, 2.0];
console.log("\n[3] Magnitudes muy dispares:");
console.log(`  JS reduce():  ${dispares.reduce((a, b) => a + b, 0)}`);
console.log(`  Neumaier Zig: ${Precision.sum(dispares)}`);
console.log(`  Esperado:     2.000000000000002`);

// [4] Varianza (suma de cuadrados centrados)
const vals = [1e8 + 1, 1e8 + 2, 1e8 + 3];
const mean = 1e8 + 2;
const sqDevs = vals.map(x => (x - mean) ** 2);
console.log("\n[4] Suma de cuadrados para varianza:");
console.log(`  Neumaier Zig: ${Precision.sum(sqDevs).toFixed(6)}  (esperado: 2.0)`);
console.log(`  JS reduce():  ${sqDevs.reduce((a, b) => a + b, 0).toFixed(6)}`);

console.log("\n[OK] Demo 02 completado.\n");

File: bindings/python/demo/02_precision.py
# Copyright (c) 2026 Edison Manrique Chocce
# Todos los derechos reservados.
# Licencia Dual: AGPL-3.0 | Comercial según LICENSE.
"""
demo/02_precision.py — Aritmética Compensada de Neumaier
Demuestra la superioridad numérica frente a la suma estándar de Python.
"""
import sys
if sys.stdout.encoding != "utf-8":
    sys.stdout.reconfigure(encoding="utf-8")

sys.path.insert(0, "..")
from ermc import Precision

print("=" * 60)
print("  ERMC | Demo 02: Aritmetica Compensada (Neumaier)")
print("=" * 60)

# --- Experimento 1: Cancelación catastrófica clásica ---
print("\n[1] Cancelacion catastrofica clasica:")
data = [1e16, 1.0, -1e16]
print(f"  Datos: {data}")
print(f"  Python sum(): {sum(data)}   <-- ERROR: devuelve 0.0")
print(f"  Neumaier Zig: {Precision.sum(data)}  <-- CORRECTO: 1.0")

# --- Experimento 2: Suma de serie armónica (1/1 + 1/2 + ... + 1/N) ---
print("\n[2] Serie armonica H_N = sum(1/k, k=1..N):")
N = 10_000
harmonic = [1.0 / k for k in range(1, N + 1)]
py_sum = sum(harmonic)
zig_sum = Precision.sum(harmonic)
# Valor de referencia: digamma(N+1) + gamma_euler
import math
ref = math.log(N) + 0.5772156649015328 + 1 / (2 * N)
print(f"  N = {N}")
print(f"  Python sum():    {py_sum:.10f}")
print(f"  Neumaier Zig:    {zig_sum:.10f}")
print(f"  Ref (log+gamma): {ref:.10f}")
print(f"  Error Python:    {abs(py_sum - ref):.2e}")
print(f"  Error Neumaier:  {abs(zig_sum - ref):.2e}")

# --- Experimento 3: Suma con magnitudes muy distintas ---
print("\n[3] Suma de magnitudes dispares [1e-15, 1e15, 1e-15, -1e15, 2.0]:")
dispares = [1e-15, 1e15, 1e-15, -1e15, 2.0]
print(f"  Python sum(): {sum(dispares)}")
print(f"  Neumaier Zig: {Precision.sum(dispares)}")
print(f"  Esperado:     2.000000000000002")

# --- Experimento 4: Compensación en cálculo de varianza ---
print("\n[4] Suma de cuadrados para varianza (Welford vs naive):")
vals = [1e8 + 1.0, 1e8 + 2.0, 1e8 + 3.0]
mean = 1e8 + 2.0
sq_devs = [(x - mean) ** 2 for x in vals]
print(f"  Suma de (x-mu)^2 por Neumaier: {Precision.sum(sq_devs):.6f}  (esperado: 2.0)")
naive_sq = sum(sq_devs)
print(f"  Suma naive:                    {naive_sq:.6f}")

print("\n[OK] Demo 02 completado.\n")
